### PR TITLE
Port addcdiv operator from the TH code to Aten

### PR DESCRIPTION
**Pull request by ifedan**
#23683 

BC: enforce stronger shape requirements on the output tensor (out= keyword argument) and do not allow output tensor to be resized if it is also used as one of the inputs.

### CODE REVIEW

**Review by VitalyFedyunin — APPROVED** [review]

File: aten/src/THC/generic/THCTensorMathPointwise.cu [hunk 1]
```diff
@@ -634,28 +634,5 @@ void THCTensor_(addcmul)(THCState *state, THCTensor *self_, THCTensor *t, scalar
...
-  }
-  else
-  {
-    THArgCheck(THCTensor_(nElement)(state, self_) == THCTensor_(nElement)(state, src1),
-               1, "sizes do not match");
-  }
-  THArgCheck(THCTensor_(nElement)(state, src1) == THCTensor_(nElement)(state, src2),
-             3, "sizes do not match");
-
-  if (!THC_pointwiseApply3<scalar_t, scalar_t, scalar_t>(state, self_, src1, src2, TensorAddCDivOp<scalar_t>(value))) {
```
Comment (fmassa) [inline aten/src/THC/generic/THCTensorMathPointwise.cu]: Cant you also remove `TensorAddCDivOp`?

Comment (ifedan) [reply aten/src/THC/generic/THCTensorMathPointwise.cu]: Done

**Comment by ifedan** [comment]
```python
import timeit
import torch
import itertools
import statistics

def test_perf(sizes, device, repeat, times):
    def _tensor(name, sizes, device):
        return '''{0} = torch.rand({1}, device="{2}");'''.format(name, sizes, device)

    setup_code = 'import torch;' + _tensor('t', sizes, device) + _tensor('t1', sizes, device) + _tensor('t2', sizes, device)
    test_code = '''torch.addcdiv(t, 0.1, t1, t2);'''
    if device == "cuda":
        test_code = test_code + 'torch.cuda.synchronize()'
    result = timeit.repeat(setup = setup_code,stmt = test_code,repeat = repeat,number = times)
    mean = statistics.mean(result)
    std = statistics.stdev(result)
    print('''sizes = {0} mean = {1} std = {2}'''.format(sizes, mean, std))

def test_perf_for_device(device, small, mid, large):
    print(device)
    for s in itertools.product((small, mid, large), (small, mid, large)):
        test_perf(str(s), device, 3, 300)

test_perf_for_device("cpu", 5, 100, 1000)
test_perf_for_device("cuda", 5, 100, 10000)
```

**pytorch:master**
```
cpu
sizes = (5, 5) mean = 0.0028569360341255865 std = 0.0010510590172268684
sizes = (5, 100) mean = 0.0028643139715616903 std = 4.3665856726755906e-05
sizes = (5, 1000) mean = 0.006352228966231148 std = 0.001127419496386336
sizes = (100, 5) mean = 0.002509646311712762 std = 0.00019963476484621904
sizes = (100, 100) mean = 0.011277683700124422 std = 0.000629327241584998
sizes = (100, 1000) mean = 0.14763347529030094 std = 0.05060114217968295
sizes = (1000, 5) mean = 0.006261350975061457 std = 0.00014816209197861848
sizes = (1000, 100) mean = 0.20003363598758975 std = 0.06528509858861585
sizes = (1000, 1000) mean = 0.12846002731627473 std = 0.049256516871917004
cuda
sizes = (5, 5) mean = 0.014807269326411188 std = 0.0002584212001880077
sizes = (5, 100) mean = 0.014873317636859914 std = 0.0005126256601174947
sizes = (5, 10000) mean = 0.014773712687504789 std = 0.00010225472887115569
sizes = (100, 5) mean = 0.014462866354733706 std = 4.4714108469530666e-05
sizes = (100, 100) mean = 0.01469070635115107 std = 5.825614780870974e-05
sizes = (100, 10000) mean = 0.042430503332676985 std = 0.0018348270532736717
sizes = (10000, 5) mean = 0.010801804989265898 std = 4.9346497464795667e-05
sizes = (10000, 100) mean = 0.044252945381837584 std = 0.00509836582659358
sizes = (10000, 10000) mean = 3.2386255230133734 std = 0.04498384965557289
```
**ifedan:addcdiv**
```
cpu
sizes = (5, 5) mean = 0.0016963026719167829 std = 0.0009315952112804344
sizes = (5, 100) mean = 0.0012403366854414344 std = 2.982969117211696e-05
sizes = (5, 1000) mean = 0.0021153349662199616 std = 1.9056974955160416e-05
sizes = (100, 5) mean = 0.0011701589683070779 std = 2.766453959914256e-05
sizes = (100, 100) mean = 0.0033482072952513895 std = 7.385991056756898e-05
sizes = (100, 1000) mean = 0.006428317011644443 std = 0.0007618354148070603
sizes = (1000, 5) mean = 0.0030287040087083974 std = 1.6911185213152017e-05
sizes = (1000, 100) mean = 0.005964546697214246 std = 3.4688374678133e-05
sizes = (1000, 1000) mean = 0.021453748689964414 std = 0.0003999432557428698
cuda
sizes = (5, 5) mean = 0.008046998642385006 std = 0.00015058991310921044
sizes = (5, 100) mean = 0.0077388930367305875 std = 7.21774730668918e-05
sizes = (5, 10000) mean = 0.007702110335230827 std = 9.980142759618565e-05
sizes = (100, 5) mean = 0.007639275281690061 std = 0.00011222829623894688
sizes = (100, 100) mean = 0.007553003340338667 std = 0.00012253531803228777
sizes = (100, 10000) mean = 0.027757083686689537 std = 7.478564008069888e-05
sizes = (10000, 5) mean = 0.007714152685366571 std = 9.202822330029497e-05
sizes = (10000, 100) mean = 0.028339064641234774 std = 0.0004282684295644353
sizes = (10000, 10000) mean = 2.095506311316664 std = 0.0005945752669524009
```

File: aten/src/ATen/native/cuda/PointwiseOpsKernel.cu [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+#include <ATen/native/PointwiseOps.h>
+#include <THC/THCNumerics.cuh>
+
+namespace at { namespace native {
+
+void addcdiv_cuda_kernel(TensorIterator& iter, Scalar value) {
+  AT_DISPATCH_ALL_TYPES_AND(kHalf, iter.dtype(), "addcdiv_cuda", [&]() {
+    auto alpha = value.to<scalar_t>();
+    gpu_kernel(iter, [alpha]GPU_LAMBDA(scalar_t a, scalar_t b, scalar_t c) -> scalar_t {
+      return THCNumerics<scalar_t>::add(a, THCNumerics<scalar_t>::mul(alpha,THCNumerics<scalar_t>::div(b, c)));
```
Comment (fmassa) [inline aten/src/ATen/native/cuda/PointwiseOpsKernel.cu]: Do we still need to use THCNumerics? In https://github.com/pytorch/pytorch/blob/7f130c8494d235054917a7abfcbe1058a65d628a/aten/src/ATen/native/cuda/BinaryOpsKernel.cu#L19
We don't use it anymore.

Comment (ifedan) [reply aten/src/ATen/native/cuda/PointwiseOpsKernel.cu]: Fixed

**Review by VitalyFedyunin — CHANGES_REQUESTED** [review]
Please add tests for integral types or disable them in dispatch

**Comment by ifedan** [comment]
> Please add tests for integral types or disable them in dispatch

Added

**Review by VitalyFedyunin — DISMISSED** [review]

**Comment by ezyang** [comment]
This diff fails lint

**Comment by VitalyFedyunin** [comment]
enforce stronger shape requirements on the output tensor (out= keyword argument) and do not allow output tensor to be resized if it is also used as one of the inputs.

**Pull request closed without merging**